test/e2e/framework/bugs.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/onsi/ginkgo/v2/types"
+)
+
+var (
+	bugs     []Bug
+	bugMutex sync.Mutex
+)
+
+// RecordBug stores information about a bug in the E2E suite source code that
+// cannot be reported through ginkgo.Fail because it was found outside of some
+// test, for example during test registration.
+//
+// This can be used instead of raising a panic. Then all bugs can be reported
+// together instead of failing after the first one.
+func RecordBug(bug Bug) {
```

Comment:
I personally don't like this behavior, on CIs that runs thousands of jobs per day this can hide important bugs or go unnoticed ... a panic makes it clear something is wrong.

> Then all bugs can be reported  together instead of failing after the first one.

How many of these cases did we have? I prefer to have more evidence that this is a real problem before adding more compelxity to the framework

Reply:
> I personally don't like this behavior, on CIs that runs thousands of jobs per day this can hide important bugs or go unnoticed ... a panic makes it clear something is wrong.

You didn't like the panic before ("panicing may cause some developer bad UX"). Also, why should this go unnoticed? It's going to be a test failure in a blocking job (more than one, actually) and each manual invocation of the E2E suite. This will be impossible to miss.

> How many of these cases did we have?

All of the white space issues in the first commit, and more once I start enforcing that inline `[]` must also be tags. You were right, panicking is a bad UX: run E2E suite, fix the one issue which caused the panic, run again, and so on.

Of course this is only during the transition and one could argue that the only developer who is going to suffer is me :sweat_smile: Later, other developers will probably only see a panic while they develop a new test and do not register it correctly. However, I am not sure how visible a panic will be in a Prow job. It might get hidden by the Ginkgo CLI which then only reports "workers failed to start".

Reply:
I checked, ginkgo reports the panic. But the UX is definitely worse. This is with `RecordBug` raising a panic instead of the current code and the whitespace commit reverted:
```console
$ _output/bin/ginkgo -p ./test/e2e
Ginkgo timed out waiting for all parallel procs to report back
Test suite: e2e (./test/e2e)

This occurs if a parallel process exits before it reports its results to the
Ginkgo CLI.  The CLI will now print out all the stdout/stderr output it's
collected from the running processes.  However you may not see anything useful
in these logs because the individual test processes usually intercept output to
stdout/stderr in order to capture it in the spec reports.

You may want to try rerunning your test suite with
--output-interceptor-mode=none to see additional output here and debug your
suite.
  
Output from proc 1:
  panic: trailing or leading spaces are unnecessary and need to be removed: "[sig-storage] vcp at scale [Feature:vsphere] "

  goroutine 1 [running]:
  k8s.io/kubernetes/test/e2e/framework.RecordBug({{0x801ddff, 0x4c}, 0x3b, {0xc000bf8900, 0x72}})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/bugs.go:43 +0x54
  k8s.io/kubernetes/test/e2e/framework.registerInSuite(0x6cc4808, {0xc000bd6f30?, 0x2d?}, {0xc0008e4aa0?, 0x2, 0xc000cfa9c0?})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:247 +0xb14
  k8s.io/kubernetes/test/e2e/framework.SIGDescribe.func1({0x6a76de6, 0x1f}, {0xc000cfa9c0, 0x1, 0x1})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:141 +0x246
  k8s.io/kubernetes/test/e2e/storage/vsphere.init()
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_scale.go:59 +0x1af

Output from proc 2:
  panic: trailing or leading spaces are unnecessary and need to be removed: "[sig-storage] vcp at scale [Feature:vsphere] "
...
Output from proc 35:
  panic: trailing or leading spaces are unnecessary and need to be removed: "[sig-storage] vcp at scale [Feature:vsphere] "

  goroutine 1 [running]:
  k8s.io/kubernetes/test/e2e/framework.RecordBug({{0x801ddff, 0x4c}, 0x3b, {0xc0001abe00, 0x72}})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/bugs.go:43 +0x54
  k8s.io/kubernetes/test/e2e/framework.registerInSuite(0x6cc4808, {0xc0006d05d0?, 0x2d?}, {0xc00055b340?, 0x2, 0xc000a919d0?})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:247 +0xb14
  k8s.io/kubernetes/test/e2e/framework.SIGDescribe.func1({0x6a76de6, 0x1f}, {0xc000a919d0, 0x1, 0x1})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:141 +0x246
  k8s.io/kubernetes/test/e2e/storage/vsphere.init()
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_scale.go:59 +0x1af

** End **
Ginkgo ran 1 suite in 22.389893721s

Test Suite Failed
```

Reply:
>  panic: trailing or leading spaces are unnecessary and need to be removed

I can't remember the exact context, but I mean, panic is used when someting went unexpectedly wrong and we are using it here to warn about "linting" on title, sounds like a big hammer

> Of course this is only during the transition and one could argue that the only developer who is going to suffer is me 😅

can't we make this more subtle? if is only for the transition ...

Reply:
A developer may always do something wrong also after the transition. If we make error reporting less obvious (for example, an error log message), then there is a considerable risk of regressions that will go unnoticed.

Is that what you meant with "subtle"?

I think the E2E suite should fail to run if test registration is not correct. I prefer the current approach with `RecordBug` because (as you said yourself :grin:) `panic` is not user-friendly. But if you find the current approach too complex, then I'm also fine with `panic`.

Reply:
With the new `ginkgo.PreviewSpecs`, bugs can be detected before even allowing the test suite to start. I think that's nicer and prevents wasting resources on testing a flawed PR, so I've switched to that.

Reply:
sgtm